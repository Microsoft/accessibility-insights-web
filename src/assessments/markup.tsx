// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import * as React from 'react';

export function Tag(props: { tagName: string; isBold?: boolean }): JSX.Element {
    const isBold = props.hasOwnProperty('isBold') ? props.isBold : true;
    return isBold ? <CodeTerm>&lt;{props.tagName}&gt;</CodeTerm> : <Code>&lt;{props.tagName}&gt;</Code>;
}

export function Emphasis(props: { children: React.ReactNode }): JSX.Element {
    return <i>{props.children}</i>;
}

export function Term(props: { children: React.ReactNode }): JSX.Element {
    return <b>{props.children}</b>;
}

export function Code(props: { children: React.ReactNode }): JSX.Element {
    return <span className="insights-code">{props.children}</span>;
}

export const GreaterThanOrEqualTo = () => <span>&#8805;</span>;

export const NonBreakingSpace = () => <span>&nbsp;</span>;

export function CodeTerm(props: { children: React.ReactNode }): JSX.Element {
    return (
        <Term>
            <Code>{props.children}</Code>
        </Term>
    );
}
