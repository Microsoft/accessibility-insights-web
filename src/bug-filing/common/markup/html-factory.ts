// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { escape } from 'lodash';
import { MarkupFactory } from './markup-factory';
import { truncateSnippet as truncate } from './truncate-snippet';

export const createFactory = (truncateSnippet: (text: string) => string): MarkupFactory => {
    const bold = (text: string): string => {
        return `<b>${text}</b>`;
    };

    const snippet = (text: string): string => {
        const truncated = truncateSnippet(text);

        return `<code>${escape(truncated)}</code>`;
    };

    const link = (href: string, text?: string): string => {
        return `<a href="${href}">${text || href}</a>`;
    };

    const sectionSeparator = () => {
        return '<br><br>';
    };

    const newLine = () => {
        return '<br>';
    };

    const howToFixSection = (failureSummary: string): string => {
        return escape(failureSummary)
            .replace(/\n  /g, '<br>- ')
            .replace(/\n /g, '<br> ')
            .replace(/\n/g, '<br>');
    };

    return {
        bold,
        snippet,
        link,
        sectionSeparator,
        newLine,
        howToFixSection,
    };
};

export const HTMLFactory = createFactory(truncate);
