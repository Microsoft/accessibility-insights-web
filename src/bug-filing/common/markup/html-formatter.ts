// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { escape } from 'lodash';
import { MarkupFormatter } from './markup-formatter';
import { truncateSnippet as truncate } from './truncate-snippet';

export const createFormatter = (truncateSnippet: (text: string) => string): MarkupFormatter => {
    const bold = (text: string): string => {
        return `<b>${text}</b>`;
    };

    const snippet = (text: string): string => {
        const truncated = truncateSnippet(text);

        return `<code>${escape(truncated)}</code>`;
    };

    const link = (href: string, text?: string): string => {
        return `<a href="${href}">${text || href}</a>`;
    };

    const sectionSeparator = () => {
        return '<br><br>';
    };

    const newLine = () => {
        return '<br>';
    };

    const howToFixSection = (failureSummary: string): string => {
        return escape(failureSummary)
            .replace(/\n  /g, '<br>- ')
            .replace(/\n /g, '<br> ')
            .replace(/\n/g, '<br>');
    };

    return {
        bold,
        snippet,
        link,
        sectionSeparator,
        newLine,
        howToFixSection,
    };
};

export const HTMLFormatter = createFormatter(truncate);
