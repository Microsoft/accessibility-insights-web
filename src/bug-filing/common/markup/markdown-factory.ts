// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { MarkupFactory } from './markup-factory';
import { truncateSnippet as truncate } from './truncate-snippet';

export const createFactory = (truncateSnippet: (text: string) => string): MarkupFactory => {
    const bold = (text: string): string => {
        return `**${text}**`;
    };

    const snippet = (text: string): string => {
        const truncated = truncateSnippet(text);

        return `\`${truncated}\``;
    };

    const link = (href: string, text?: string): string => {
        return `[${text || href}](${href})`;
    };

    const sectionSeparator = (): string => {
        return '';
    };

    const newLine = (): string => {
        return '';
    };

    const howToFixSection = (failureSummary: string): string => {
        return failureSummary
            .split('\n')
            .map(line => `    ${line}`)
            .join('\n');
    };

    return {
        bold,
        snippet,
        link,
        sectionSeparator,
        newLine,
        howToFixSection,
    };
};

export const MarkdownFactory = createFactory(truncate);
