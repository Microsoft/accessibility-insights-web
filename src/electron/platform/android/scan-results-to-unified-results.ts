// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { InstancePropertyBag, InstanceResultStatus, UnifiedResult } from 'common/types/store-data/unified-data-interface';
import { UUIDGeneratorType } from 'common/uid-generator';
import { DictionaryStringTo } from '../../../types/common-types';
import { RuleInformation } from './rule-information';
import { RuleInformationProvider } from './rule-information-provider';
import { RuleResultsData, ScanResults, ViewElementData } from './scan-results';

export type ConvertScanResultsToUnifiedResultsDelegate = (scanResults: ScanResults, uuidGenerator: UUIDGeneratorType) => UnifiedResult[];

export function convertScanResultsToUnifiedResults(scanResults: ScanResults, uuidGenerator: UUIDGeneratorType): UnifiedResult[] {
    if (!scanResults || !scanResults.ruleResults) {
        return [];
    }

    return createUnifiedResultsFromScanResults(scanResults, uuidGenerator);
}

function createUnifiedResultsFromScanResults(scanResults: ScanResults, uuidGenerator: UUIDGeneratorType): UnifiedResult[] {
    const ruleInformationProvider: RuleInformationProvider = new RuleInformationProvider();
    const viewElementLookup: DictionaryStringTo<ViewElementData> = createViewElementLookup(scanResults);
    const unifiedResults: UnifiedResult[] = [];

    for (const ruleResult of scanResults.ruleResults) {
        const ruleInformation: RuleInformation = ruleInformationProvider.getRuleInformation(ruleResult.ruleId);

        if (ruleInformation) {
            unifiedResults.push(createUnifiedResult(ruleInformation, ruleResult, viewElementLookup, uuidGenerator));
        }
    }

    return unifiedResults;
}

function createViewElementLookup(scanResults: ScanResults): DictionaryStringTo<ViewElementData> {
    const viewElementLookup = {};

    addViewElementAndChildren(viewElementLookup, scanResults.viewElementTree);

    return viewElementLookup;
}

function addViewElementAndChildren(viewElementLookup: DictionaryStringTo<ViewElementData>, element: ViewElementData): void {
    if (element) {
        viewElementLookup[element.axeViewId] = element;
        if (element.children) {
            for (const child of element.children) {
                addViewElementAndChildren(viewElementLookup, child);
            }
        }
    }
}

function createUnifiedResult(
    ruleInformation: RuleInformation,
    ruleResult: RuleResultsData,
    viewElementLookup: DictionaryStringTo<ViewElementData>,
    uuidGenerator: UUIDGeneratorType,
): UnifiedResult {
    return {
        uid: uuidGenerator(),
        ruleId: ruleInformation.ruleId,
        status: getStatus(ruleResult.status),
        descriptors: getDescriptors(viewElementLookup[ruleResult.axeViewId]),
        identifiers: null,
        resolution: {
            'how-to-fix': ruleInformation.howToFix(ruleResult),
        },
    };
}

function getDescriptors(viewElement: ViewElementData): InstancePropertyBag {
    if (viewElement) {
        return {
            className: viewElement.className,
            boundingRectangle: viewElement.boundsInScreen,
            contentDescription: viewElement.contentDescription,
            text: viewElement.text,
        };
    }
    return null;
}

function getStatus(status: string): InstanceResultStatus {
    switch (status) {
        case 'PASS':
            return 'pass';
        case 'FAIL':
            return 'fail';
        default:
            return 'unknown';
    }
}
