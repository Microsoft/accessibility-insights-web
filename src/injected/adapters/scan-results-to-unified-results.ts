// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { flatMap } from 'lodash';

import { InstanceResultStatus, UnifiedResult } from '../../common/types/store-data/unified-data-interface';
import { UUIDGeneratorType } from '../../common/uid-generator';
import { IssueFilingUrlStringUtils } from '../../issue-filing/common/issue-filing-url-string-utils';
import { AxeNodeResult, RuleResult, ScanResults } from '../../scanner/iruleresults';

export type ConvertScanResultsToUnifiedResultsDelegate = (scanResults: ScanResults, uuidGenerator: UUIDGeneratorType) => UnifiedResult[];

interface RuleResultData {
    status: InstanceResultStatus;
    ruleID: string;
}

interface CreationData extends RuleResultData {
    cssSelector: string;
    snippet: string;
    howToFix: {
        oneOf: string[];
        none: string[];
        all: string[];
    };
}

export const convertScanResultsToUnifiedResults = (scanResults: ScanResults, uuidGenerator: UUIDGeneratorType): UnifiedResult[] => {
    if (!scanResults) {
        return [];
    }
    return createUnifiedResultsFromScanResults(scanResults, uuidGenerator);
};

const createUnifiedResultsFromScanResults = (scanResults: ScanResults, uuidGenerator: UUIDGeneratorType): UnifiedResult[] => {
    return [
        ...createUnifiedResultsFromRuleResults(scanResults.violations, 'fail', uuidGenerator),
        ...createUnifiedResultsFromRuleResults(scanResults.passes, 'pass', uuidGenerator),
    ];
};

const createUnifiedResultsFromRuleResults = (
    ruleResults: RuleResult[],
    status: InstanceResultStatus,
    uuidGenerator: UUIDGeneratorType,
): UnifiedResult[] => {
    const unifiedResultFromRuleResults = (ruleResults || []).map(result =>
        createUnifiedResultsFromRuleResult(result, status, uuidGenerator),
    );

    return flatMap(unifiedResultFromRuleResults);
};

const createUnifiedResultsFromRuleResult = (
    ruleResult: RuleResult,
    status: InstanceResultStatus,
    uuidGenerator: UUIDGeneratorType,
): UnifiedResult[] => {
    return ruleResult.nodes.map(node => {
        const data: RuleResultData = {
            status: status,
            ruleID: ruleResult.id,
        };

        return createUnifiedResultFromNode(node, data, uuidGenerator);
    });
};

const createUnifiedResultFromNode = (
    nodeResult: AxeNodeResult,
    ruleResultData: RuleResultData,
    uuidGenerator: UUIDGeneratorType,
): UnifiedResult => {
    return createUnifiedResult(
        {
            ...ruleResultData,
            cssSelector: nodeResult.target.join(';'),
            snippet: nodeResult.snippet || nodeResult.html,
            howToFix: {
                oneOf: nodeResult.any.map(checkResult => checkResult.message),
                none: nodeResult.none.map(checkResult => checkResult.message),
                all: nodeResult.all.map(checkResult => checkResult.message),
            },
        },
        uuidGenerator,
    );
};

const createUnifiedResult = (data: CreationData, uuidGenerator: UUIDGeneratorType): UnifiedResult => {
    return {
        uid: uuidGenerator(),
        status: data.status,
        ruleId: data.ruleID,
        identifiers: {
            'css-selector': data.cssSelector,
            shortName: IssueFilingUrlStringUtils.getSelectorLastPart(data.cssSelector),
        },
        descriptors: {
            snippet: data.snippet,
        },
        resolution: {
            'how-to-fix-web': {
                any: data.howToFix.oneOf,
                none: data.howToFix.none,
                all: data.howToFix.all,
            },
        },
    } as UnifiedResult;
};
