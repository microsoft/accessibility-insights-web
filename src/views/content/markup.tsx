// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { Code, Emphasis, Tag, Term } from 'assessments/markup';
import { NewTabLink } from 'common/components/new-tab-link';
import { CheckIcon } from 'common/icons/check-icon';
import { CrossIcon } from 'common/icons/cross-icon';
import { ContentActionMessageCreator } from 'common/message-creators/content-action-message-creator';
import { TextContent } from 'content/strings/text-content';
import * as React from 'react';
import { Helmet } from 'react-helmet';
import { CodeExample, CodeExampleProps } from './markup/code-example';

type PassFailProps = {
    passText: JSX.Element;
    passExample?: React.ReactNode;
    failText: JSX.Element;
    failExample?: React.ReactNode;
};

export type Markup = {
    Tag: React.FC;
    Code: React.FC;
    Term: React.FC;
    Emphasis: React.FC;
    Do: React.FC;
    Dont: React.FC;
    Pass: React.FC;
    Fail: React.FC;
    PassFail: React.FC<PassFailProps>;
    Columns: React.FC;
    Column: React.FC;
    Inline: React.FC;
    HyperLink: React.FC<{ href: string }>;
    Title: React.FC<{ children: string }>;
    Highlight: React.FC;
    CodeExample: React.FC<CodeExampleProps>;
    Links: React.FC;
    Table: React.FC;
    LandmarkLegend: React.FC<{ role: string }>;
    ProblemList: React.FC;
    Include: React.FC<{ content: MarkupBasedComponent }>;
};

export type MarkupDeps = {
    textContent: Pick<TextContent, 'applicationTitle'>;
    contentActionMessageCreator: Pick<ContentActionMessageCreator, 'openContentHyperLink'>;
};

export interface MarkupOptions {
    setPageTitle?: boolean;
}

export type MarkupBasedComponentProps = {
    deps: MarkupDeps;
    options?: MarkupOptions;
};

export type MarkupBasedComponent = React.FC<MarkupBasedComponentProps>;

export const createMarkup = (deps: MarkupDeps, options?: MarkupOptions) => {
    function Include(props: { content: MarkupBasedComponent }): JSX.Element {
        const Content = props.content;
        return <Content deps={deps} options={options} />;
    }

    function Title(props: { children: string }): JSX.Element {
        const { applicationTitle } = deps.textContent;
        const helmet = (
            <Helmet>
                <title>
                    {props.children} - {applicationTitle}
                </title>
            </Helmet>
        );

        return (
            <>
                {options && options.setPageTitle && helmet}
                <h1>{props.children}</h1>
            </>
        );
    }

    function HyperLink(props: { href: string; children: React.ReactNode }): JSX.Element {
        const { href } = props;
        const { openContentHyperLink } = deps.contentActionMessageCreator;

        return (
            <NewTabLink href={href} onClick={e => openContentHyperLink(e, href)}>
                {props.children}
            </NewTabLink>
        );
    }

    function Links(props: { children: React.ReactNode }): JSX.Element {
        return (
            <>
                <div className="content-hyperlinks">
                    {React.Children.map(props.children, el => el)}
                </div>
            </>
        );
    }

    function Inline(props: { children: React.ReactNode }): JSX.Element {
        return <div className="content-inline">{props.children}</div>;
    }

    function Do(props: { children: React.ReactNode }): JSX.Element {
        return (
            <Column>
                <div className="do-header">
                    <h2>Do</h2>
                    <CheckIcon />
                </div>
                <div className="do-section">{props.children}</div>
            </Column>
        );
    }

    function Dont(props: { children: React.ReactNode }): JSX.Element {
        return (
            <Column>
                <div className="dont-header">
                    <h2>Don't</h2>
                    <CrossIcon />
                </div>
                <div className="dont-section">{props.children}</div>
            </Column>
        );
    }

    function Pass(props: { children: React.ReactNode }): JSX.Element {
        return (
            <Column>
                <div className="pass-header">
                    <CheckIcon /> <h3>Pass</h3>
                </div>
                <div className="pass-section">{props.children}</div>
            </Column>
        );
    }

    function Fail(props: { children: React.ReactNode }): JSX.Element {
        return (
            <Column>
                <div className="fail-header">
                    <CrossIcon /> <h3>Fail</h3>
                </div>
                <div className="fail-section">{props.children}</div>
            </Column>
        );
    }

    function LandmarkLegend(props: { role: string; children: React.ReactNode }): JSX.Element {
        return <span className={`landmarks-legend ${props.role}-landmark`}>{props.children}</span>;
    }

    function Highlight(props: { children: React.ReactNode }): JSX.Element {
        return <span className="highlight">{props.children}</span>;
    }

    function Table(props: { children: React.ReactNode }): JSX.Element {
        return <ul className="table">{props.children}</ul>;
    }

    function ProblemList(props: { children: React.ReactNode }): JSX.Element {
        return <ul className="accessibility-problems-list">{props.children}</ul>;
    }

    function Columns(props: { children: React.ReactNode }): JSX.Element {
        return <div className="columns">{props.children}</div>;
    }

    function Column(props: { children: React.ReactNode }): JSX.Element {
        return <div className="column">{props.children}</div>;
    }

    function PassFail(props: PassFailProps): React.ReactNode {
        const { passText, passExample, failText, failExample } = props;

        function formatExample(example: string | React.ReactNode): React.ReactNode {
            if (typeof example === 'string') {
                return <CodeExample>{example}</CodeExample>;
            }
            return example;
        }

        return (
            <div className="pass-fail-grid">
                <div className="fail-section">
                    <div className="fail-header">
                        <CrossIcon /> <h3>Fail</h3>
                    </div>
                    {failText}
                </div>
                {failExample && <div className="fail-example">{formatExample(failExample)}</div>}
                <div className="pass-section">
                    <div className="pass-header">
                        <CheckIcon /> <h3>Pass</h3>
                    </div>
                    {passText}
                </div>
                {passExample && <div className="pass-example">{formatExample(passExample)}</div>}
            </div>
        );
    }

    return {
        Tag,
        Code,
        Term,
        Emphasis,
        Do,
        Dont,
        Pass,
        Fail,
        PassFail,
        Columns,
        Column,
        Inline,
        HyperLink,
        Title,
        CodeExample,
        Highlight,
        Links,
        LandmarkLegend,
        Table,
        ProblemList,
        options,
        Include,
    } as Markup;
};
