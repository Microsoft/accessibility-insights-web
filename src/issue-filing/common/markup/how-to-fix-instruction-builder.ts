// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { isEmpty } from 'lodash';

import { HowToFixInstructions } from '../../../injected/adapters/scan-results-to-unified-results';

export type InstructionListBuilder = (title: string, checks: string[]) => string;
export const buildHowToFixInstructions = (instructions: HowToFixInstructions, listBuilder: InstructionListBuilder) => {
    const getInstructionList = (checks: string[], title: string) => {
        if (isEmpty(checks)) {
            return '';
        }

        const displayTitle = checks.length === 1 ? 'Fix the following: ' : title;
        return listBuilder(displayTitle, checks);
    };

    const all = instructions.all.concat(instructions.none);
    const oneOf = instructions.oneOf;
    const fixAllInstructions = getInstructionList(all, 'Fix ALL of the following: ');
    const fixAnyInstructions = getInstructionList(oneOf, 'Fix ONE of the following: ');
    return `${fixAllInstructions}${fixAnyInstructions}`;
};
