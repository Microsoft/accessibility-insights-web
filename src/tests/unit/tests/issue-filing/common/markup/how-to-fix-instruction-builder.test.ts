// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { HowToFixInstructions } from '../../../../../../injected/adapters/scan-results-to-unified-results';
import { buildHowToFixInstructions } from '../../../../../../issue-filing/common/markup/how-to-fix-instruction-builder';

describe('buildHowToFixInstructions', () => {
    const howToFixCombinations: HowToFixInstructions[] = [
        { oneOf: [], all: [], none: [] },
        { oneOf: [], all: ['all'], none: [] },
        { oneOf: [], all: ['all'], none: ['none'] },
        { oneOf: [], all: [], none: ['none'] },
        { oneOf: ['any'], all: [], none: [] },
        { oneOf: ['any'], all: ['all'], none: [] },
        { oneOf: ['any'], all: ['all'], none: ['none'] },
        { oneOf: ['any'], all: [], none: ['none'] },
        { oneOf: ['any-1', 'any-2'], all: ['all-1', 'all-2'], none: ['none-1', 'none-2'] },
    ];

    test.each(howToFixCombinations)('build instructions', instructions => {
        const result = buildHowToFixInstructions(instructions, (title, checks) => `${title}${checks.join('; ')}; `);
        expect(result).toMatchSnapshot();
    });
});
