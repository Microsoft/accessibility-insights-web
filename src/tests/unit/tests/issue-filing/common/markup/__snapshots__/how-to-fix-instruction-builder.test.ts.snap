// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildHowToFixInstructions build instructions 1`] = `""`;

exports[`buildHowToFixInstructions build instructions 2`] = `"Fix the following: all; "`;

exports[`buildHowToFixInstructions build instructions 3`] = `"Fix ALL of the following: all; none; "`;

exports[`buildHowToFixInstructions build instructions 4`] = `"Fix the following: none; "`;

exports[`buildHowToFixInstructions build instructions 5`] = `"Fix the following: any; "`;

exports[`buildHowToFixInstructions build instructions 6`] = `"Fix the following: all; Fix the following: any; "`;

exports[`buildHowToFixInstructions build instructions 7`] = `"Fix ALL of the following: all; none; Fix the following: any; "`;

exports[`buildHowToFixInstructions build instructions 8`] = `"Fix the following: none; Fix the following: any; "`;

exports[`buildHowToFixInstructions build instructions 9`] = `"Fix ALL of the following: all-1; all-2; none-1; none-2; Fix ONE of the following: any-1; any-2; "`;
